File: recursion/freecodecamp/rec1_string.c
/****************************************************************
*
* File name     : rec1_string.c
* Author        : Linda J N
* Creation date : Saturday 18 September 2021 09:17:37 PM
* Purpose       :
*
****************************************************************/
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define printfs(...) printf("%s", __VA_ARGS__)
#define printfd(...) printf("%d", __VA_ARGS__)
#define printfc(...) printf("%c", __VA_ARGS__)
#define printl printf("\n")

/* Function to print reverse of the passed string */
void reverse(char *str)
{
   if (*str != '\0')
   {
       reverse(str+1);
       printfc(*str);
   }
}

char* rev(char in[]) {
  int n = strlen(in);

  for(unsigned int i = 0; i < n/2; i++)
    in[i] = in[n - i - 1];
  return in;
}

int main(int argc, char *argv[]) {
  reverse("abc"); printl;
  return 0;
}

File: recursion/freecodecamp/rec1_string.py
#!/usr/bin/env python
################################################################
# File name     : rec1_string.py
# Purpose       :
# Author        : Linda J
# Creation date : Sunday 19 September 2021 09:04:10 AM
################################################################

import sys

def reverse_string(inp):
    n = len(inp) - 1
    if inp == '':
        return inp
    #reverse string from start till n-1, as nth one is appended
    return inp[n] + reverse_string(inp[0:n])

def main():
    print(reverse_string(""))
    print(reverse_string("a"))
    print(reverse_string("abcd"))
    print(reverse_string("abcde"))
    print(reverse_string(''))
    return 0

if __name__ == "__main__":
    main()
